### PR TITLE
Ice dynamics

### PR DESCRIPTION
**Pull request by OlgaSergienko**
This PR includes major changes that enable a new flux with ocean_only runs - the surface mass flux over Antarctica (`shelf_sfc_mass_flux`). Changes have been made to MOM_forcing_type.F90 and MOM_ice_shelf.F90.

### CODE REVIEW

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 [hunk 1]
```diff
@@ -461,8 +453,8 @@ subroutine initialize_ice_flow_from_file(bed_elev,u_shelf, v_shelf,float_cond,&
 
   floatfr_varname = "float_frac"
 
- call MOM_read_data(filename, trim(ushelf_varname), u_shelf, G%Domain, position=CORNER,scale=1.0)
- call MOM_read_data(filename,trim(vshelf_varname), v_shelf, G%Domain, position=CORNER,scale=1.0)
+! call MOM_read_data(filename, trim(ushelf_varname), u_shelf, G%Domain, position=CORNER,scale=1.0)
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_initialize.F90:456]: Please explain why these two lines are commented out, but the (now unused) variables specifying the names of these variables are still being read.
Perhaps instead of commenting these lines  out, it would make more sense to test whether there is a non-zero length variable name, and only do the call to MOM_read_data() if there is a valid string for the variable name.

File: src/core/MOM_forcing_type.F90 [hunk 2]
```diff
@@ -2910,6 +2918,10 @@ subroutine forcing_diagnostics(fluxes_in, sfc_state, G_in, US, time_end, diag, h
     if ((handles%id_ustar_ice_cover > 0) .and. associated(fluxes%ustar_shelf)) &
       call post_data(handles%id_ustar_ice_cover, fluxes%ustar_shelf, diag)
 
+    if ((handles%id_shelf_sfc_mass_flux > 0) &
```
Comment (Hallberg-NOAA) [inline src/core/MOM_forcing_type.F90:2921]: There does not appear to be a register_diag_field call anywhere for handles%id_shelf_sfc_mass_flux. Is this deliberate, or is this being handled in some other way?

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 [hunk 3]
```diff
@@ -169,7 +169,7 @@ subroutine initialize_ice_thickness_from_file(h_shelf, area_shelf_h, hmask, G, U
 
         if (area_shelf_h(i,j) >= G%areaT(i,j)) then
           hmask(i,j) = 1.
-          area_shelf_h(i,j)=G%areaT(i,j)
+!          area_shelf_h(i,j)=G%areaT(i,j)
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_initialize.F90:172]: Why would it make sense for the area covered by the shelf in a grid cell to exceed the area of the grid cell itself?

File: src/ice_shelf/MOM_ice_shelf.F90 [hunk 4]
```diff
@@ -1975,6 +2002,57 @@ subroutine initialize_shelf_mass(G, param_file, CS, ISS, new_sim)
...
+      endif
+      ISS%mass_shelf(i,j) = ISS%h_shelf(i,j) * CS%density_ice
+    endif
+  enddo ; enddo
+
+  call pass_var(ISS%area_shelf_h, G%domain)
+  call pass_var(ISS%h_shelf, G%domain)
+  call pass_var(ISS%hmask, G%domain)
+  call pass_var(ISS%mass_shelf, G%domain)
+
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf.F90:2053]: The four halo updates here would be more efficient if they were combined by adding the optional argument `complete=.false.` to the first three, and then (optionally)  `complete=.true.` to the fourth one.  Answers should be bitwise identical, but this flag would let FMS know to combine the messages, which saves on latency as the various PEs are synced up.

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 (hunk 3 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf_initialize.F90:172]: `area_shelf_f` is read from a file, `G%areaT` is computed at the initialization of the horizontal grid. The two might not be identical.

File: src/ice_shelf/MOM_ice_shelf.F90 [hunk 5]
```diff
@@ -1815,6 +1840,8 @@ subroutine initialize_ice_shelf(param_file, ocn_grid, Time, CS, diag, forces_in,
   if (CS%active_shelf_dynamics) then
     CS%id_h_mask = register_diag_field('ice_shelf_model', 'h_mask', CS%diag%axesT1, CS%Time, &
        'ice shelf thickness mask', 'none')
+    CS%id_shelf_sfc_mass_flux = register_diag_field('ice_shelf_model', 'sfc_mass_flux', CS%diag%axesT1, CS%Time, &
+       'ice shelf surface mass flux deposition from atmosphere', 'none')
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf.F90:1844]: I suspect that this register_diag_field call is missing the optional argument `conversion=conversion=US%RZ_T_to_kg_m2s`.

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 (hunk 3 above)
Comment (Hallberg-NOAA) [reply src/ice_shelf/MOM_ice_shelf_initialize.F90:172]: I think that it is fine that they are not identical, but the line that was commented out deals with this possibility by enforcing that the shelf_area_h can not exceed the area of the cell, which seems to me like a very natural restriction.  What would it mean if the ratio of the areas as (area_shelf_h/G%areaT) were greater than 1?

File: src/ice_shelf/MOM_ice_shelf.F90 (hunk 4 above)
Comment (Hallberg-NOAA) [reply src/ice_shelf/MOM_ice_shelf.F90:2053]: Yes, that is what I mean, but only for the first 3 calls.  The fourth call would have to be `call pass_var(ISS%mass_shelf, G%domain, complete=.true.)` so that the halo updates actually get done.

File: src/ice_shelf/MOM_ice_shelf.F90 [hunk 6]
```diff
@@ -321,6 +325,11 @@ subroutine shelf_calc_flux(sfc_state_in, fluxes_in, Time, time_step, CS)
   G => CS%grid ; US => CS%US
   ISS => CS%ISS
 
+  if (CS%data_override_shelf_fluxes .and. CS%active_shelf_dynamics) then
+       call data_override(G%Domain, 'shelf_sfc_mass_flux', fluxes_in%shelf_sfc_mass_flux, CS%Time, &
+                          scale=US%kg_m3_to_R*US%m_to_Z)
```
[the remark this answers is not in the document]
Comment (MJHarrison-GFDL) [reply src/ice_shelf/MOM_ice_shelf.F90:330]: Thank you. I have added the correct rescaling in this [PR](https://github.com/OlgaSergienko/MOM6/pull/10)

File: src/ice_shelf/MOM_ice_shelf.F90 (hunk 5 above)
Comment (MJHarrison-GFDL) [reply src/ice_shelf/MOM_ice_shelf.F90:1844]: Fixed in this [PR](https://github.com/OlgaSergienko/MOM6/pull/10)

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 (hunk 1 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf_initialize.F90]: These lines are uncommented.

File: src/ice_shelf/MOM_ice_shelf.F90 (hunk 4 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf.F90:2053]: These calls are corrected.

File: src/ice_shelf/MOM_ice_shelf_initialize.F90 (hunk 3 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf_initialize.F90]: The line is uncommented. The ratio of areas cannot be greater than 1.

**Comment by Hallberg-NOAA** [comment]
This PR should be ready to go once Matt's PR back to OlgaSergienko:ice_dynamics at  https://github.com/OlgaSergienko/MOM6/pull/10 is accepted.

**Comment by OlgaSergienko** [comment]
Matt's PR has been merged and all requested changes have been addressed.

**Review by Hallberg-NOAA — APPROVED** [review]

**Comment by marshallward** [comment]
@OlgaSergienko @MJHarrison-GFDL Is a squash-merge ok here?  It will wipe out your commit histories.  Otherwise, one of you will need to clean up the rebase conflicts.
